batch.go

```diff
@@ -991,3 +1004,121 @@ func (batch *Batch) Equal(other Batcher) bool {
 	}
 	return len(batch.Entries) == equalEntries && equalEntries != 0
 }
+
+// WithOffset sets the Offset information onto a Batch so that during Create a "balanced offset"
+// record is appended at the end.
+//
+// Balanced offset records sum a File's Debits or Credits to zero in a single transaction.
+// Offset files are used to offset transactions from a single account inside of the ODFI.
+//
+// If the Batch contains Debits the offset record will be a Credit and if the batch contains Credits
+// the offset record will be a debit.
+func (b *Batch) WithOffset(off *Offset) {
+	b.offset = off
+}
+
+func (b *Batch) upsertOffset() error {
+	if b == nil || b.offset == nil {
+		return nil
+	}
+
+	// remove any Offset records already on the batch
+	addOffset := true // append to EntryAddendaCount
+	for i := 0; i < len(b.Entries); i++ {
+		// TODO(adam): Should we remove this based on checking the last element is
+		// debit/credit and sums to all the other elements (which are mutually exclusive to
+		// the last record being debit or credit)?
+		if strings.EqualFold(b.Entries[i].IndividualName, "OFFSET") {
+			// fixup BatchControl records for our conditional after this for loop
+			if b.Entries[i].TransactionCode == CheckingCredit || b.Entries[i].TransactionCode == SavingsCredit {
+				b.Control.TotalCreditEntryDollarAmount = 0
+			} else {
+				b.Control.TotalDebitEntryDollarAmount = 0
+			}
+			// remove the EntryDetail
+			b.Entries = append(b.Entries[:i], b.Entries[i+i:]...)
+			addOffset = false // we're replacing an EntryDetail, not adding
+			i++
+		}
+	}
+	if b.Control.TotalDebitEntryDollarAmount > 0 && b.Control.TotalCreditEntryDollarAmount > 0 {
```

Comment:
Is this true that an offset record can only have debits or credits but not both? I thought that the customer in #general said she couldn't have both.  Wouldn't we just change the service calls code based on debit, credit, or both? 
200 ACH DEBITS AND CREDITS
220 ACH CREDITS ONLY
225 ACH DEBITS ONLY

Reply:
I think you would have to change the service class code to 200 if it was 220 or 225 and and offset record is created.  I think if you were going to create offset records for a mixed batch, would the code stop creating offsets after the first offset for credits/debits?

Reply:
I understood our chats and their problem as only allowing debits or credits within a batch and the offset was the opposite to zero out. 

If a Batch can have debits and credits and needs a balancing record that's fine. We can do the math to figure it out still.

Reply:
A batch can have debits and credits and have an offset. I think these are two problems. That customer needs to separate debits and credits into two different files with each having an offset. We have a separate feature request for segmenting debits and credits. 
https://github.com/moov-io/ach/issues/528

Reply:
So a batch with debits and credits can't have an offset record created? It sounds like in that case we should return an error and direct the client to split the file and then call `WithOffset` on each.